services/vision/obstaclesdepth/obstacles_depth.go

```diff
@@ -298,17 +294,26 @@ func (o *obsDepth) isCompatible(p1, p2 image.Point) bool {
 	return true
 }
 
-// clustersToBoxes will turn the clusters we get from kmeans into boxes.
-func (o *obsDepth) clustersToBoxes(clusters clusters.Clusters) ([]spatialmath.Geometry, error) {
+// performKMeans3D will do k-means clustering on projected obstacle points.
+func (o *obsDepth) performKMeans3D(k int) ([]spatialmath.Geometry, clusters.Clusters, error) {
+	var d clusters.Observations
+	for _, pt := range o.obstaclePts {
+		outX, outY, outZ := o.intrinsics.PixelToPoint(float64(pt.X), float64(pt.Y), float64(o.dm.GetDepth(pt.X, pt.Y)))
```

Comment:
Since pixel values are presumably always whole numbers, curious why this function accepts floats rather than taking ints and converting to floats internally. Not needed to be changed in this PR if you don't want to though.

Reply:
Nah you're making a valid point but I _think_ the main thing is the depth value (z) can be naturally floatish and maybe it's weird to have a mix between ints and floats.  To your point tho, we have a function `func intrinsics2DPtTo3DPt(pt image.Point, d rimage.Depth, pci *PinholeCameraIntrinsics) (r3.Vector, error) {` but I was thinking it's unexported for a reason.

Reply:
I'm gonna make a ticket for this lol